vault/identity_store_oidc_provider.go

```diff
@@ -277,7 +278,163 @@ func oidcProviderPaths(i *IdentityStore) []*framework.Path {
 			HelpSynopsis:    "Query OIDC configurations",
 			HelpDescription: "Query this path to retrieve the configured OIDC Issuer and Keys endpoints, response types, subject types, and signing algorithms used by the OIDC backend.",
 		},
+		{
+			Pattern: "oidc/provider/" + framework.GenericNameRegex("name") + "/.well-known/keys",
+			Fields: map[string]*framework.FieldSchema{
+				"name": {
+					Type:        framework.TypeString,
+					Description: "Name of the provider",
+				},
+			},
+			Callbacks: map[logical.Operation]framework.OperationFunc{
+				logical.ReadOperation: i.pathOIDCReadProviderPublicKeys,
+			},
+			HelpSynopsis:    "Retrieve public keys",
+			HelpDescription: "Returns the public portion of keys for a named OIDC provider. Clients can use them to validate the authenticity of an ID token.",
+		},
+	}
+}
+
+// TODO: load clients into memory (go-memdb) to look this up
+func (i *IdentityStore) allClients(ctx context.Context, s logical.Storage) (*[]client, error) {
```

Comment:
Is it possible to fetch the actual client objects through `i. clientByName()` that you added in https://github.com/hashicorp/vault/pull/12538? That way we only hit storage on cache misses.

Reply:
Indeed! That's the plan. The reason that I mentioned keeping this function around was for the case of `allowed_client_ids="*"`, where we need to list all clients. I don't think we can reliably list all clients from memdb because we can't control when a client could become invalidated. For example, on client updates the client is invalidated (i.e., deleted) in from memdb. A subsequent list wouldn't contain the full set of clients.

So I see a need to keep this list all clients from storage code around + use the new memdb `clientBy*` functions. It will depend on what's set for `allowed_client_ids`.

Reply:
Good points, thanks!

@calvn Yesterday Austin and I agreed to review this PR as-is and incorporate the changes from https://github.com/hashicorp/vault/pull/12538 in another PR. I can update the PR description to reflect that.